<a href="https://colab.research.google.com/github/cassidyhanna/DS-Unit-1-Sprint-2-Data-Wrangling-and-Storytelling/blob/master/DS-Unit-1-Sprint-2-Data-Wrangling-and-Storytelling/S2_M1_LS_DS_121_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Lambda School Data Science_

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [2]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-09-20 18:16:31--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.106.214
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.106.214|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  71.1MB/s    in 2.8s    

2019-09-20 18:16:34 (71.1 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [3]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [4]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [5]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

Dataframes
* orders.csv
* products.csv

Columns
- product name
- add to cart order

In [0]:
##### YOUR CODE HERE #####
import pandas as pd
import numpy as np
import seaborn as sns


# aisles  

In [7]:
aisles = pd.read_csv("aisles.csv")
aisles.head()

aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

In [8]:
aisles.shape

(134, 2)

# **departments**

In [9]:
departments = pd.read_csv("departments.csv")
departments.head()

department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [10]:
departments.shape

(21, 2)

# **Orders**


In [11]:
orders = pd.read_csv("orders.csv")
orders.head()

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [12]:
orders.shape

(3421083, 7)

# **products**

In [13]:
products = pd.read_csv("products.csv")
products.head()

product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [14]:
products.shape

(49688, 4)

In [0]:
product_name = products['product_name']

In [16]:
product_name.head()

0                           Chocolate Sandwich Cookies
1                                     All-Seasons Salt
2                 Robust Golden Unsweetened Oolong Tea
3    Smart Ones Classic Favorites Mini Rigatoni Wit...
4                            Green Chile Anytime Sauce
Name: product_name, dtype: object

In [17]:
product_name.shape

(49688,)

This tells us theres 49688 different products

# **order_products_prior**

In [18]:
order_products_prior = pd.read_csv("order_products__prior.csv")
order_products_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [19]:
order_products_prior.shape

(32434489, 4)

starting to think i should compare product ID to add_to_cart_order

# **order train**

In [20]:
order_train = pd.read_csv("order_products__train.csv")
order_train.head()

order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [21]:
order_train.shape

(1384617, 4)

- product_id of order_train 

- add_to_cart_order of order_train 

## Concatenate order_products__prior and order_products__train

In [22]:
order_products = pd.concat([order_products_prior,order_train])
order_products.shape

(33819106, 4)

In [23]:
order_products.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

# **Grouping**

In [24]:
order_products.groupby('order_id')['product_id'].count().mean()

10.10707325550502

In [0]:
grouped_orders = order_products.groupby('order_id')

In [26]:
grouped_orders.get_group(2539329)

order_id  product_id  add_to_cart_order  reordered
24076664   2539329         196                  1          0
24076665   2539329       14084                  2          0
24076666   2539329       12427                  3          0
24076667   2539329       26088                  4          0
24076668   2539329       26405                  5          0

In [27]:
order_products[order_products['order_id']==2539329]

order_id  product_id  add_to_cart_order  reordered
24076664   2539329         196                  1          0
24076665   2539329       14084                  2          0
24076666   2539329       12427                  3          0
24076667   2539329       26088                  4          0
24076668   2539329       26405                  5          0

In [28]:
grouped_orders['product_id'].count()

order_id
1           8
2           9
3           8
4          13
5          26
6           3
7           2
8           1
9          15
10         15
11          5
12         15
13         13
14         11
15          5
16          3
18         28
19          3
20          8
21          5
22         14
23         14
24          3
25         14
26          8
27         27
28         16
29          5
30          3
31         10
           ..
3421053     9
3421055    19
3421056     5
3421057     5
3421058     8
3421059     6
3421060    17
3421061    22
3421062     7
3421063     4
3421064     3
3421065     5
3421066     6
3421067     1
3421068    14
3421069    12
3421070     3
3421071     5
3421072    12
3421073     2
3421074     4
3421075     8
3421076     8
3421077     4
3421078     9
3421079     1
3421080     9
3421081     7
3421082     7
3421083    10
Name: product_id, Length: 3346083, dtype: int64

# get a subset of orders 

In [29]:
orders.head()

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [30]:
orders.shape

(3421083, 7)

In [0]:
condition = (orders['user_id'] == 1) & (orders['order_number'] <= 2)
columns = ['order_id', 'user_id', 'order_number','order_dow','order_hour_of_day']
subset = orders[condition][columns]

In [32]:
subset.head()

order_id  user_id  order_number  order_dow  order_hour_of_day
0   2539329        1             1          2                  8
1   2398795        1             2          3                  7

# **Merge Dataframes**

In [33]:
columns = ['order_id','product_id','add_to_cart_order']
merged = pd.merge(subset,order_products[columns])
merged.head()

order_id  user_id  ...  product_id  add_to_cart_order
0   2539329        1  ...         196                  1
1   2539329        1  ...       14084                  2
2   2539329        1  ...       12427                  3
3   2539329        1  ...       26088                  4
4   2539329        1  ...       26405                  5

[5 rows x 7 columns]

merge with columns from product

In [34]:
final = pd.merge(merged, products[['product_id', 'product_name']])
final

order_id  ...                             product_name
0    2539329  ...                                     Soda
1    2398795  ...                                     Soda
2    2539329  ...  Organic Unsweetened Vanilla Almond Milk
3    2539329  ...                      Original Beef Jerky
4    2398795  ...                      Original Beef Jerky
5    2539329  ...               Aged White Cheddar Popcorn
6    2398795  ...               Aged White Cheddar Popcorn
7    2539329  ...         XL Pick-A-Size Paper Towel Rolls
8    2398795  ...                               Pistachios
9    2398795  ...                   Bag of Organic Bananas
10   2398795  ...                    Cinnamon Toast Crunch

[11 rows x 8 columns]

In [0]:
id_ = order_products[columns]

In [36]:
id_.head()

order_id  product_id  add_to_cart_order
0         2       33120                  1
1         2       28985                  2
2         2        9327                  3
3         2       45918                  4
4         2       30035                  5

In [0]:
main = pd.merge(id_, products[['product_id', 'product_name']])

In [38]:
main.head()

order_id  product_id  add_to_cart_order        product_name
0         2       33120                  1  Organic Egg Whites
1        26       33120                  5  Organic Egg Whites
2       120       33120                 13  Organic Egg Whites
3       327       33120                  5  Organic Egg Whites
4       390       33120                 28  Organic Egg Whites

In [39]:
main.shape

(33819106, 4)

In [0]:
product_name = main['product_name'] 

In [0]:
topten = main.product_name.value_counts()[:10]

In [42]:
topten 

Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

In [0]:
df = pd.DataFrame(topten) 

In [44]:
df

product_name
Banana                        491291
Bag of Organic Bananas        394930
Organic Strawberries          275577
Organic Baby Spinach          251705
Organic Hass Avocado          220877
Organic Avocado               184224
Large Lemon                   160792
Strawberries                  149445
Limes                         146660
Organic Whole Milk            142813

In [45]:
df.shape

(10, 1)

In [46]:
main.product_name.unique()[:10]

array(['Organic Egg Whites', 'Michigan Organic Kale', 'Garlic Powder',
       'Coconut Butter', 'Natural Sweetener', 'Carrots',
       'Original Unflavored Gelatine Mix',
       'All Natural No Stir Creamy Almond Butter',
       'Classic Blend Cole Slaw',
       'Total 2% with Strawberry Lowfat Greek Strained Yogurt'],
      dtype=object)

Banana

Bag of Organic Bananas

Organic Strawberries

Organic Baby Spinach

Organic Hass Avocado

Organic Avocado

Large Lemon

Strawberries

Limes

Organic Whole Milk

## Reshape Data Section

- Replicate the lesson code
- Complete the code cells we skipped near the beginning of the notebook
- Table 2 --> Tidy
- Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [0]:
table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

In [48]:
table1

treatmenta  treatmentb
John Smith           NaN           2
Jane Doe            16.0          11
Mary Johnson         3.0           1

In [57]:
table1.index

Index(['John Smith', 'Jane Doe', 'Mary Johnson'], dtype='object')

In [0]:
table1 = table1.reset_index()

In [60]:
table1

index  treatmenta  treatmentb
0    John Smith         NaN           2
1      Jane Doe        16.0          11
2  Mary Johnson         3.0           1

In [62]:
tidy = table1.melt(id_vars='index')
tidy

index    variable  value
0    John Smith  treatmenta    NaN
1      Jane Doe  treatmenta   16.0
2  Mary Johnson  treatmenta    3.0
3    John Smith  treatmentb    2.0
4      Jane Doe  treatmentb   11.0
5  Mary Johnson  treatmentb    1.0

In [63]:
tidy.columns = ['name','trt','result']
tidy

name         trt  result
0    John Smith  treatmenta     NaN
1      Jane Doe  treatmenta    16.0
2  Mary Johnson  treatmenta     3.0
3    John Smith  treatmentb     2.0
4      Jane Doe  treatmentb    11.0
5  Mary Johnson  treatmentb     1.0

In [49]:
table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [64]:
table2.index

Index(['treatmenta', 'treatmentb'], dtype='object')

In [0]:
table2 = table2.reset_index()

In [68]:
tidy1 = table2.melt(id_vars='index')
tidy1

index      variable  value
0  treatmenta    John Smith    NaN
1  treatmentb    John Smith    2.0
2  treatmenta      Jane Doe   16.0
3  treatmentb      Jane Doe   11.0
4  treatmenta  Mary Johnson    3.0
5  treatmentb  Mary Johnson    1.0

In [69]:
tidy1.columns = ['trt', 'name', 'result']
tidy1

trt          name  result
0  treatmenta    John Smith     NaN
1  treatmentb    John Smith     2.0
2  treatmenta      Jane Doe    16.0
3  treatmentb      Jane Doe    11.0
4  treatmenta  Mary Johnson     3.0
5  treatmentb  Mary Johnson     1.0

In [0]:
table2

In [71]:
tidy1.pivot_table(index = 'trt', columns = 'name', values = 'result')

name        Jane Doe  John Smith  Mary Johnson
trt                                           
treatmenta      16.0         NaN           3.0
treatmentb      11.0         2.0           1.0

In [0]:
flights = sns.load_dataset('flights')

In [54]:
##### YOUR CODE HERE #####
flights.head()

year     month  passengers
0  1949   January         112
1  1949  February         118
2  1949     March         132
3  1949     April         129
4  1949       May         121

In [0]:
flight_table = pd.pivot_table(flights, index=['year'] , columns=['month'])

In [56]:
flight_table

passengers                       ...                                    
month    January February March April  ... September October November December
year                                   ...                                    
1949         112      118   132   129  ...       136     119      104      118
1950         115      126   141   135  ...       158     133      114      140
1951         145      150   178   163  ...       184     162      146      166
1952         171      180   193   181  ...       209     191      172      194
1953         196      196   236   235  ...       237     211      180      201
1954         204      188   235   227  ...       259     229      203      229
1955         242      233   267   269  ...       312     274      237      278
1956         284      277   317   313  ...       355     306      271      306
1957         315      301   356   348  ...       404     347      305      336
1958         340      318   362   348  ...       404     359      310      337
1959         360      342   406   396  ...       463     407      362      405
1960         417      391   419   461  ...       508     461      390      432

[12 rows x 12 columns]

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####